### PR TITLE
docs: Add debug console to dev guide

### LINKED ISSUE #72: Explain how to setup a debug console
Whilst we wait for https://github.com/kata-containers/agent/issues/169 to finish simmering, we need to document in the developer guide how to manually setup a debug console for debugging purposes.

### PR DESCRIPTION
**Pull request by jodh-intel**
Add an appendix explaining how to setup a debug console to login to the
virtual machine for debugging.

Fixes #72.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

### CODE REVIEW

**Comment by jodh-intel** [comment]
Hi @egernst, @devimc, @klynnrif - ptal.

File: Developer-Guide.md [hunk 1]
```diff
@@ -24,6 +24,15 @@
 * [Troubleshoot Kata Containers](#troubleshoot-kata-containers)
 * [Appendices](#appendices)
     * [Checking Docker default runtime](#checking-docker-default-runtime)
+    * [Setting up a debug console](#setting-up-a-debug-console)
```
Comment (klynnrif) [inline Developer-Guide.md]: Suggest changing this title to "Set up a debug console" to avoid using the gerund (ing) format.

File: Developer-Guide.md [hunk 2]
```diff
@@ -277,10 +286,131 @@ To perform analysis on Kata logs, use the
 [`kata-log-parser`](https://github.com/kata-containers/tests/tree/master/cmd/log-parser)
 tool.
 
+See also [Setting up a debug console](#setting-up-a-debug-console).
```
Comment (klynnrif) [inline Developer-Guide.md]: Suggest removing "also": See [Setting up a debug console](#setting-up-a-debug-console).

File: Developer-Guide.md [hunk 3]
```diff
@@ -277,10 +286,131 @@ To perform analysis on Kata logs, use the
...
 
 ## Checking Docker default runtime
 
 ```
 $ sudo docker info 2>/dev/null | grep -i "default runtime" | cut -d: -f2- | grep -q runc  && echo "SUCCESS" || echo "ERROR: Incorrect default Docker runtime"
 ```
+
+## Setting up a debug console
+
+By default, it is not possible to login to a virtual machine since this could
```
Comment (klynnrif) [inline Developer-Guide.md]: Lines 301-304 suggested rewrite (unless I've changed the meaning): By default you cannot login to a virtual machine since this can be sensitive from a security perspective. However a virtual machine also requires additional packages in the rootfs, which means the image used to boot each virtual machine is larger.

File: Developer-Guide.md [hunk 4]
```diff
@@ -277,10 +286,131 @@ To perform analysis on Kata logs, use the
...
+be sensitive from a security perspective. However it also requires additional
+packages in the rootfs which means the image used to boot each virtual machine
+would be larger.
+
+If you wish to be able to login to a virtual machine that hosts your
```
Comment (klynnrif) [inline Developer-Guide.md]: Lines 306-307 suggested rewrite: If you want to login to a virtual machine that hosts your containers, complete the following steps, which assume a rootfs image.

File: Developer-Guide.md [hunk 5]
```diff
@@ -277,10 +286,131 @@ To perform analysis on Kata logs, use the
...
+containers, follow the steps below which assume a rootfs image.
+
+### Create a custom image containing a shell
+
+To be able to login to a virtual machine, you must [create a custom
```
Comment (klynnrif) [inline Developer-Guide.md]: To login to a virtual machine...

File: Developer-Guide.md [hunk 6]
```diff
@@ -277,10 +286,131 @@ To perform analysis on Kata logs, use the
...
+
+```
+$ cd $GOPATH/src/github.com/kata-containers/osbuilder/rootfs-builder
+$ export ROOTFS_DIR=${GOPATH}/src/github.com/kata-containers/osbuilder/rootfs-builder/rootfs
+$ script -fec 'sudo -E GOPATH=$GOPATH USE_DOCKER=true EXTRA_PKGS="bash" ./rootfs.sh centos'
+```
+
+### Create a debug systemd service
+
+Create the service file that will start the shell in the rootfs directory:
```
Comment (klynnrif) [inline Developer-Guide.md]: Create the service file that starts the shell in the rootfs directory:

File: Developer-Guide.md [hunk 7]
```diff
@@ -277,10 +286,131 @@ To perform analysis on Kata logs, use the
...
+Environment=PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
+StandardInput=tty
+StandardOutput=tty
+PrivateDevices=yes
+Type=simple
+ExecStart=/usr/bin/bash
+EOT
+```
+
+**Note**: You may need to adjust the `ExecStart=` path.
```
Comment (klynnrif) [inline Developer-Guide.md]: **Note**: You might need to adjust the `ExecStart=` path.

File: Developer-Guide.md [hunk 8]
```diff
@@ -277,10 +286,131 @@ To perform analysis on Kata logs, use the
...
+
+```
+$ name="kata-containers-centos-with-debug-console.img"
+$ sudo install -o root -g root -m 0640 kata-containers.img "/usr/share/kata-containers/${name}"
+```
+
+Next, modify the `image=` values in the `[hypervisor.qemu]` section of the
+[configuration file](https://github.com/kata-containers/runtime#configuration)
+to specify the full path to the image name specified in the previous code
+section. Alternatively, recreate the symbolic link so that it is pointing to
```
Comment (klynnrif) [inline Developer-Guide.md]: ...Alternatively, recreate the symbolic link so it points to

File: Developer-Guide.md [hunk 9]
```diff
@@ -277,10 +286,131 @@ To perform analysis on Kata logs, use the
...
+the new debug image:
+
+```
+$ (cd /usr/share/kata-containers && sudo ln -sf "$name" kata-containers.img)
+```
+
+**Note**: You should take care to undo this change after finishing debugging
```
Comment (klynnrif) [inline Developer-Guide.md]: **Note**: You should take care to undo this change after you finish debugging

File: Developer-Guide.md [hunk 10]
```diff
@@ -277,10 +286,131 @@ To perform analysis on Kata logs, use the
...
+```
+
+**Note**: You need to press the `RETURN` key to see the shell prompt.
+
+To disconnect from the virtual machine, type `CONTROL+q` (hold down the
+`CONTROL` key and press `q`).
+
+### Obtain details of the image
+
+If the image was created using
```
Comment (klynnrif) [inline Developer-Guide.md]: If the image is created using

File: Developer-Guide.md [hunk 11]
```diff
@@ -277,10 +286,131 @@ To perform analysis on Kata logs, use the
...
+[osbuilder](https://github.com/kata-containers/osbuilder), the following YAML
+file will exist and contain details of the image and how it was created:
```
Comment (klynnrif) [inline Developer-Guide.md]: file exists and contains details of the image and how it was created:

File: Developer-Guide.md (hunk 3 above)
Comment (jodh-intel) [reply Developer-Guide.md]: After a coffee, that paragraph reads rather strangely in both forms :)

I've tried to reword using your suggested phrasing and come up with:

> By default you cannot login to a virtual machine since this can be sensitive
from a security perspective. Also allowing logins would require additional
packages in the rootfs, which would increase the size of the image used to
boot the virtual machine.

**Comment by jodh-intel** [comment]
Thanks @klynnrif - branch updated.

File: Developer-Guide.md (hunk 3 above)
Comment (klynnrif) [reply Developer-Guide.md]: Ha! I like your phrasing. Thanks :)

**Comment by jodh-intel** [comment]
ping @kata-containers/documentation.

File: Developer-Guide.md [hunk 12]
```diff
@@ -277,10 +286,132 @@ To perform analysis on Kata logs, use the
 [`kata-log-parser`](https://github.com/kata-containers/tests/tree/master/cmd/log-parser)
 tool.
 
+See Set up a debug console](#set-up-a-debug-console).
```
Comment (caoruidong) [inline Developer-Guide.md]: lost a [ ?

Comment (jodh-intel) [reply Developer-Guide.md]: Good catch @caoruidong! ;) Branch updated.

**Comment by jodh-intel** [comment]
Hi @egernst, @bergwolf - ptal.
